File: lib/src/domain/rsi_calculator.dart
/// RSI Calculator — Pure domain logic.
///
/// Computes the Relative Strength Index (14-period by default) using
/// Wilder's smoothing (exponential smoothed average of gains/losses).
/// Standard formula:
///   RSI = 100 - 100 / (1 + RS)  where RS = avg_gain / avg_loss
library;

import 'entities.dart';

class RsiCalculator {
  const RsiCalculator();

  /// Compute the current RSI for the given [period] (default 14).
  ///
  /// Returns null if fewer than [period + 1] candles are available.
  double? compute(List<DailyCandle> candles, {int period = 14}) {
    final series = computeSeries(candles, period: period);
    for (int i = series.length - 1; i >= 0; i--) {
      if (series[i].$2 != null) return series[i].$2;
    }
    return null;
  }

  /// Compute a rolling RSI series aligned with [candles].
  ///
  /// The first [period] entries will have null values (warmup).
  List<(DateTime, double?)> computeSeries(
    List<DailyCandle> candles, {
    int period = 14,
  }) {
    if (candles.length <= period) {
      return candles.map((c) => (c.date, null as double?)).toList();
    }

    final result = <(DateTime, double?)>[];
    // Seed with average gain/loss of first [period] price changes
    double sumGain = 0;
    double sumLoss = 0;
    for (int i = 1; i <= period; i++) {
      final change = candles[i].close - candles[i - 1].close;
      if (change >= 0) {
        sumGain += change;
      } else {
        sumLoss -= change; // keep positive
      }
    }
    double avgGain = sumGain / period;
    double avgLoss = sumLoss / period;

    // Nulls for warmup
    for (int i = 0; i < period; i++) {
      result.add((candles[i].date, null));
    }

    // First RSI
    result.add((candles[period].date, _rsi(avgGain, avgLoss)));

    // Rolling Wilder smoothing
    for (int i = period + 1; i < candles.length; i++) {
      final change = candles[i].close - candles[i - 1].close;
      final gain = change > 0 ? change : 0.0;
      final loss = change < 0 ? -change : 0.0;
      avgGain = (avgGain * (period - 1) + gain) / period;
      avgLoss = (avgLoss * (period - 1) + loss) / period;
      result.add((candles[i].date, _rsi(avgGain, avgLoss)));
    }

    return result;
  }

  static double _rsi(double avgGain, double avgLoss) {
    if (avgLoss == 0) return 100.0;
    final rs = avgGain / avgLoss;
    return 100.0 - 100.0 / (1.0 + rs);
  }
}

File: src/domain/rsi-calculator.ts
/**
 * RSI Calculator — Pure domain logic.
 * Ported from Dart: lib/src/domain/rsi_calculator.dart
 *
 * RSI = 100 - 100 / (1 + RS), where RS = avgGain / avgLoss (Wilder smoothing).
 */
import type { DailyCandle } from "../types/domain";
import { DEFAULTS } from "./technical-defaults";

export interface RsiPoint {
  readonly date: string;
  readonly value: number | null;
}

function rsi(avgGain: number, avgLoss: number): number {
  if (avgLoss === 0) return 100.0;
  const rs = avgGain / avgLoss;
  return 100.0 - 100.0 / (1.0 + rs);
}

/**
 * Compute a rolling RSI series aligned with candles.
 * The first `period` entries have null values (warmup).
 */
export function computeRsiSeries(
  candles: readonly DailyCandle[],
  period = DEFAULTS.period,
): RsiPoint[] {
  if (candles.length <= period) {
    return candles.map((c) => ({ date: c.date, value: null }));
  }

  const result: RsiPoint[] = [];

  // Seed: average gain/loss of first `period` price changes
  let sumGain = 0;
  let sumLoss = 0;
  for (let i = 1; i <= period; i++) {
    const prev = candles[i - 1];
    const curr = candles[i];
    if (!prev || !curr) continue;
    const change = curr.close - prev.close;
    if (change >= 0) {
      sumGain += change;
    } else {
      sumLoss -= change;
    }
  }
  let avgGain = sumGain / period;
  let avgLoss = sumLoss / period;

  // Null warmup
  for (let i = 0; i < period; i++) {
    const candle = candles[i];
    if (candle) {
      result.push({ date: candle.date, value: null });
    }
  }

  // First RSI
  const firstCandle = candles[period];
  if (firstCandle) {
    result.push({ date: firstCandle.date, value: rsi(avgGain, avgLoss) });
  }

  // Rolling Wilder smoothing
  for (let i = period + 1; i < candles.length; i++) {
    const prev = candles[i - 1];
    const curr = candles[i];
    if (!prev || !curr) continue;
    const change = curr.close - prev.close;
    const gain = change > 0 ? change : 0.0;
    const loss = change < 0 ? -change : 0.0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    result.push({ date: curr.date, value: rsi(avgGain, avgLoss) });
  }

  return result;
}

/** Compute the current (latest) RSI value. Returns null if insufficient data. */
export function computeRsi(
  candles: readonly DailyCandle[],
  period = DEFAULTS.period,
): number | null {
  const series = computeRsiSeries(candles, period);
  for (let i = series.length - 1; i >= 0; i--) {
    const point = series[i];
    if (point?.value !== null && point?.value !== undefined) return point.value;
  }
  return null;
}
